src/libpmemfile/preload.c

```diff
@@ -1348,15 +1349,39 @@ log_write(const char *fmt, ...)
 	syscall_no_intercept(SYS_write, log_fd, buf, len);
 }
 
-static void
+/*
+ * open_new_pool -- attempts to open a pmemfile_pool
+ * Initializes the fields called pool and pmem_stat in a pool_description
+ * struct. Does nothing if they are already initialized. The most
+ * important part of this initialization is of course calling
+ * pmemfile_pool_open.
+ *
+ * Returns:
+ *  On success: the same pointer as the argument
+ *  If the pool was already open: the same pointer as the argument
+ *  On failure: NULL pointer
+ */
+static struct pool_description *
 open_new_pool(struct pool_description *p)
 {
+	struct pool_description *ret = p;
+
+	util_mutex_lock(&p->pool_open_lock);
+
 	if (p->pool == NULL) {
```

Comment:
BTW, I think technically, that variable should be `volatile` for this. In pre C99 / pre C++11 memory model the compiler can freely add suprious store anywhere it wants to, e.g. generating this is completely legal, since it doesn't affect the program's output (as long as the standard doesn't know about threads):
```
mov [pool], %rax
mov $123, [pool]
mov %rax, [pool]
```